src/lib.rs

```diff
@@ -1,19 +1,38 @@
+#![no_std]
+
+extern crate alloc;
+
+#[doc(hidden)]
+pub use alloc::string::String;
+
 #[macro_export]
 macro_rules! string_concat {
-    ( $( $x:expr ),* ) => {
-        {
-            let mut temp_string = String::new();
-            $(
-                temp_string.push_str($x);
-            )*
-            temp_string
-        }
+    ( $( $x:expr ),* ) => {{
+        $crate::string_concat_impl!{[] $($x),*}
+    }};
+}
+
+#[doc(hidden)]
+#[macro_export]
+macro_rules! string_concat_impl {
+    ([$($x:ident)*] ) => {{
+        let mut temp_string = $crate::String::with_capacity(0$(+$x.len())*);
+        $(
+            temp_string.push_str($x);
+        )*
+        temp_string
+    }};
+    ([$($ident:ident)*] $x:expr $(, $rest:expr)*) => {
```

Comment:
This is a macro hygiene trick, where you can define an arbitrary number of local identifiers.